packages/horizon/contracts/data-service/utilities/ProvisionManager.sol

```diff
@@ -0,0 +1,352 @@
+ * - Provision tokens
+ */
+abstract contract ProvisionManager is Initializable, GraphDirectory, ProvisionManagerV1Storage {
+    using UintRange for uint256;
+
+    /**
+     * @notice Emitted when the provision tokens range is set.
+     * @param min The minimum allowed value for the provision tokens.
+     * @param max The maximum allowed value for the provision tokens.
+     */
+    event ProvisionTokensRangeSet(uint256 min, uint256 max);
+
+    /**
+     * @notice Emitted when the delegation ratio range is set.
+     * @param min The minimum allowed value for the delegation ratio.
+     * @param max The maximum allowed value for the delegation ratio.
+     */
+    event DelegationRatioRangeSet(uint32 min, uint32 max);
+
+    /**
+     * @notice Emitted when the verifier cut range is set.
+     * @param min The minimum allowed value for the max verifier cut.
+     * @param max The maximum allowed value for the max verifier cut.
+     */
+    event VerifierCutRangeSet(uint32 min, uint32 max);
+
+    /**
+     * @notice Emitted when the thawing period range is set.
+     * @param min The minimum allowed value for the thawing period.
+     * @param max The maximum allowed value for the thawing period.
+     */
+    event ThawingPeriodRangeSet(uint64 min, uint64 max);
+
+    /**
+     * @notice Thrown when a provision parameter is out of range.
+     * @param message The error message.
+     * @param value The value that is out of range.
+     * @param min The minimum allowed value.
+     * @param max The maximum allowed value.
+     */
+    error ProvisionManagerInvalidValue(bytes message, uint256 value, uint256 min, uint256 max);
+
+    /**
+     * @notice Thrown when the caller is not authorized to manage the provision of a service provider.
+     * @param caller The address of the caller.
+     * @param serviceProvider The address of the service provider.
+     */
+    error ProvisionManagerNotAuthorized(address caller, address serviceProvider);
+
+    /**
+     * @notice Thrown when a provision is not found.
+     * @param serviceProvider The address of the service provider.
+     */
+    error ProvisionManagerProvisionNotFound(address serviceProvider);
+
+    /**
+     * @notice Checks if the caller is authorized to manage the provision of a service provider.
+     */
+    modifier onlyProvisionAuthorized(address serviceProvider) {
+        require(
+            _graphStaking().isAuthorized(msg.sender, serviceProvider, address(this)),
+            ProvisionManagerNotAuthorized(msg.sender, serviceProvider)
+        );
+        _;
+    }
+
+    /**
+     * @notice Checks if a provision of a service provider is valid according
+     * to the parameter ranges established.
+     */
+    modifier onlyValidProvision(address serviceProvider) virtual {
+        IHorizonStaking.Provision memory provision = _getProvision(serviceProvider);
+        _checkProvisionTokens(provision);
+        _checkProvisionDelegationRatio(provision, serviceProvider);
```

Comment:
Since we're not checking this in SubgraphService, I kinda think we should remove the delegation ratio check from here as well... I'd expect other data services to use delegation ratio in the same way as SubgraphService, using the ratio when computing available stake